pyat/at/physics/rdt.py

```diff
@@ -0,0 +1,545 @@
+
+
+__all__ = ["get_rdts", "RDTType"]
+
+
+class RDTType(Enum):
+    """Enum class for RDT type"""
+
+    ALL = 0
+    FOCUSING = 1
+    COUPLING = 2
+    CHROMATIC = 3
+    GEOMETRIC1 = 4
+    GEOMETRIC2 = 5
+    TUNESHIFT = 6
+
+
+@dataclass
+class _RDT:
+    # Location
+    refpts: Sequence[int] | int = None
+    # Normal quadrupoles rdts
+    h20000: Sequence[complex] | complex = None
+    h00200: Sequence[complex] | complex = None
+    # linear coupling rdts
+    h10010: Sequence[complex] | complex = None
+    h10100: Sequence[complex] | complex = None
+    # chromatic rdts
+    h11001: Sequence[complex] | complex = None
+    h00111: Sequence[complex] | complex = None
+    h20001: Sequence[complex] | complex = None
+    h00201: Sequence[complex] | complex = None
+    h10002: Sequence[complex] | complex = None
+    # sextupole geometric rdts
+    h21000: Sequence[complex] | complex = None
+    h30000: Sequence[complex] | complex = None
+    h10110: Sequence[complex] | complex = None
+    h10020: Sequence[complex] | complex = None
+    h10200: Sequence[complex] | complex = None
+    # octupole geometric rdts
+    h22000: Sequence[complex] | complex = None
+    h11110: Sequence[complex] | complex = None
+    h00220: Sequence[complex] | complex = None
+    h31000: Sequence[complex] | complex = None
+    h40000: Sequence[complex] | complex = None
+    h20110: Sequence[complex] | complex = None
+    h11200: Sequence[complex] | complex = None
+    h20020: Sequence[complex] | complex = None
+    h20200: Sequence[complex] | complex = None
+    h00310: Sequence[complex] | complex = None
+    h00400: Sequence[complex] | complex = None
+    # Detuning
+    dnux_dJx: Sequence[float] | float = None
+    dnux_dJy: Sequence[float] | float = None
+    dnuy_dJy: Sequence[float] | float = None
+
+    def __getattr__(self, item):
+        return asdict(self)[item]
+
+    def __getitem__(self, item):
+        return asdict(self)[item]
+
+
+def _get_polynom(elem, attr, index):
+    try:
+        val = getattr(elem, attr)[index]
+        return val
+    except (IndexError, AttributeError):
+        return 0
+
+
+def _computedrivingterms(
+    s,
+    betax,
+    betay,
+    phix,
+    phiy,
+    etax,
+    a2l,
+    b2l,
+    b3l,
+    b4l,
+    tune,
+    rdttype,
+    nperiods,
+    refpt,
+    second_order,
+):
+    """
+    Original implementation from ELEGANT
+    Based on J.Bengtsson, SLS Note 9 / 97, March 7, 1997, with corrections per W.Guo (NSLS)
+    Revised to follow C.X.Wang AOP - TN - 2009 - 020 for second - order terms
+    """
+    periodicfactor = numpy.ones((9, 9))
+    rdts = {}
+
+    def pf(i, j):
+        return periodicfactor[4 + i][4 + j]
+
+    if nperiods != 1:
+        for i in range(9):
+            for j in range(9):
+                a1 = numpy.pi * 2 * (tune[0] * (i - 4) + tune[1] * (j - 4))
+                a2 = a1 / nperiods
+                periodicfactor[i][j] = (numpy.exp(1j * a1) - 1.0) / (
```

Comment:
Also, this expression is invalid for $x=1$. You get a `numpy.nan`  inside the `periodicfactor` array. See here below
```python
array([[ 0.11995293,  0.71065224,  1.14536311,  0.86937818,  0.02918188,
        -0.51789102,  0.09712676,  1.92673149,  4.00214385],
       [ 0.8729706 ,  1.15079497,  0.67337023, -0.18950072, -0.4889994 ,
         0.49218752,  2.5152446 ,  4.42114406,  4.96390501],
       [ 1.09715717,  0.44600387, -0.36645857, -0.36237996,  0.96769714,
         3.095958  ,  4.73701569,  4.8011494 ,  3.24475565],
       [ 0.20445964, -0.48205199, -0.13425928,  1.50518507,  3.64058563,
         4.93338244,  4.51554829,  2.67085531,  0.61087161],
       [-0.51931852,  0.19301704,  2.08165682,  4.12193798,         **nan**,
         4.12193798,  2.08165682,  0.19301704, -0.51931852],
       [ 0.61087161,  2.67085531,  4.51554829,  4.93338244,  3.64058563,
         1.50518507, -0.13425928, -0.48205199,  0.20445964],
       [ 3.24475565,  4.8011494 ,  4.73701569,  3.095958  ,  0.96769714,
        -0.36237996, -0.36645857,  0.44600387,  1.09715717],
       [ 4.96390501,  4.42114406,  2.5152446 ,  0.49218752, -0.4889994 ,
        -0.18950072,  0.67337023,  1.15079497,  0.8729706 ],
       [ 4.00214385,  1.92673149,  0.09712676, -0.51789102,  0.02918188,
         0.86937818,  1.14536311,  0.71065224,  0.11995293]])
```

Reply:
`periodicfactor` is defined by the machine tune and does not depend on the driving term.
 It is now calculated again per driving term.
Consider to move it up in the function call to avoid repeating the same complex calculation at the expense of an extra variable of size 9x9 numpy.complex128 memory cost.

Reply:
Done, `x=1` correspond to a `tune=0.0` you will get an unstable ring error before entering this function